File: backends/java/zappy-adapters/src/main/java/com/zappymart/adapters/graphql/AuthenticationPayload.java
package com.zappymart.adapters.graphql;

import com.zappymart.domain.accounts.Customer;
import com.zappymart.domain.shared.UserError;

import java.time.Instant;
import java.util.List;

public record AuthenticationPayload(
        Customer customer,
        String accessToken,
        Instant accessTokenExpiresAt,
        List<UserError> errors) {

    public static AuthenticationPayload refused(List<UserError> errors) {
        return new AuthenticationPayload(null, null, null, errors);
    }
}

File: backends/dotnet/src/Zappy.Adapters.GraphQL/Payloads/AuthenticationPayload.cs
using Zappy.Application;
using Zappy.Domain;

namespace Zappy.Adapters.GraphQL;

public sealed class AuthenticationPayload(
    Customer? customer,
    string? accessToken,
    DateTimeOffset? accessTokenExpiresAt,
    IReadOnlyList<UserError> errors)
{
    public Customer? Customer { get; } = customer;

    public string? AccessToken { get; } = accessToken;

    public DateTimeOffset? AccessTokenExpiresAt { get; } = accessTokenExpiresAt;

    public IReadOnlyList<UserError> Errors { get; } = errors;

    public static AuthenticationPayload From(Zappy.Domain.Result<Authentication> result) =>
        result.Value is null
            ? new AuthenticationPayload(null, null, null, result.Errors)
            : new AuthenticationPayload(
                result.Value.Customer,
                result.Value.AccessToken,
                result.Value.AccessTokenExpiresAt,
                result.Errors);
}
